File: cpp/src/Ice/LocalExceptions.cpp
// Copyright (c) ZeroC, Inc.

#include "Ice/LocalExceptions.h"
#include "Ice/Initialize.h"
#include "Ice/Proxy.h"
#include "Ice/StringUtil.h"
#include "Network.h"

#include <sstream>

using namespace std;
using namespace Ice;
using namespace IceInternal;

namespace
{
    string createDispatchExceptionMessage(ReplyStatus replyStatus, optional<string> message)
    {
        if (message)
        {
            return std::move(*message);
        }
        else
        {
            ostringstream os;
            os << "The dispatch failed with reply status " << replyStatus << '.';
            return os.str();
        }
    }

    optional<string> createRequestFailedMessage(
        Ice::ReplyStatus replyStatus,
        const Identity& id,
        const string& facet,
        const string& operation)
    {
        if (id.name.empty())
        {
            return nullopt;
        }
        else
        {
            std::ostringstream os;
            os << "Dispatch failed with " << replyStatus << " { id = '" << identityToString(id, ToStringMode::Unicode)
               << "', facet = '" << facet << "', operation = '" << operation << "' }";
            return os.str();
        }
    }
}

Ice::DispatchException::DispatchException(const char* file, int line, ReplyStatus replyStatus, optional<string> message)
    : LocalException{file, line, createDispatchExceptionMessage(replyStatus, std::move(message))},
      _replyStatus{replyStatus}
{
    if (_replyStatus <= ReplyStatus::UserException)
    {
        throw std::invalid_argument{
            "the replyStatus of a DispatchException must be greater than ReplyStatus::UserException"};
    }
}

const char*
Ice::DispatchException::ice_id() const noexcept
{
    return "::Ice::DispatchException";
}

Ice::RequestFailedException::~RequestFailedException() = default; // avoid weak vtable.

Ice::RequestFailedException::RequestFailedException(
    const char* file,
    int line,
    ReplyStatus replyStatus,
    Identity id,
    string facet,
    string operation)
    : DispatchException(file, line, replyStatus, createRequestFailedMessage(replyStatus, id, facet, operation)),
      _id{make_shared<Identity>(std::move(id))},
      _facet{make_shared<string>(std::move(facet))},
      _operation{make_shared<string>(std::move(operation))}
{
}

Ice::RequestFailedException::RequestFailedException(const char* file, int line, ReplyStatus replyStatus)
    : DispatchException(file, line, replyStatus),
      _id{make_shared<Identity>()},
      _facet{make_shared<string>()},
      _operation{make_shared<string>()}
{
}

Ice::ObjectNotExistException::ObjectNotExistException(
    const char* file,
    int line,
    Identity id,
    string facet,
    string operation)
    : RequestFailedException(
          file,
          line,
          ReplyStatus::ObjectNotExist,
          std::move(id),
          std::move(facet),
          std::move(operation))
{
}

Ice::ObjectNotExistException::ObjectNotExistException(const char* file, int line)
    : RequestFailedException(file, line, ReplyStatus::ObjectNotExist)
{
}

const char*
Ice::ObjectNotExistException::ice_id() const noexcept
{
    return "::Ice::ObjectNotExistException";
}

Ice::FacetNotExistException::FacetNotExistException(
    const char* file,
    int line,
    Identity id,
    string facet,
    string operation)
    : RequestFailedException(
          file,
          line,
          ReplyStatus::FacetNotExist,
          std::move(id),
          std::move(facet),
          std::move(operation))
{
}

Ice::FacetNotExistException::FacetNotExistException(const char* file, int line)
    : RequestFailedException(file, line, ReplyStatus::FacetNotExist)
{
}

const char*
Ice::FacetNotExistException::ice_id() const noexcept
{
    return "::Ice::FacetNotExistException";
}

Ice::OperationNotExistException::OperationNotExistException(
    const char* file,
    int line,
    Identity id,
    string facet,
    string operation)
    : RequestFailedException(
          file,
          line,
          ReplyStatus::OperationNotExist,
          std::move(id),
          std::move(facet),
          std::move(operation))
{
}

Ice::OperationNotExistException::OperationNotExistException(const char* file, int line)
    : RequestFailedException(file, line, ReplyStatus::OperationNotExist)
{
}

const char*
Ice::OperationNotExistException::ice_id() const noexcept
{
    return "::Ice::OperationNotExistException";
}

Ice::UnknownException::UnknownException(const char* file, int line, string message)
    : DispatchException{file, line, ReplyStatus::UnknownException, std::move(message)}
{
}

const char*
Ice::UnknownException::ice_id() const noexcept
{
    return "::Ice::UnknownException";
}

Ice::UnknownException::UnknownException(const char* file, int line, ReplyStatus replyStatus, string message)
    : DispatchException{file, line, replyStatus, std::move(message)}
{
}

Ice::UnknownLocalException::UnknownLocalException(const char* file, int line, string message)
    : UnknownException{file, line, ReplyStatus::UnknownLocalException, std::move(message)}
{
}

const char*
Ice::UnknownLocalException::ice_id() const noexcept
{
    return "::Ice::UnknownLocalException";
}

Ice::UnknownUserException::UnknownUserException(const char* file, int line, string message)
    : UnknownException{file, line, ReplyStatus::UnknownUserException, std::move(message)}
{
}

UnknownUserException
Ice::UnknownUserException::fromTypeId(const char* file, int line, const char* typeId)
{
    ostringstream os;
    os << "the reply carries a user exception that does not conform to the operation's exception specification: ";
    os << typeId;
    return UnknownUserException{file, line, os.str()};
}

const char*
Ice::UnknownUserException::ice_id() const noexcept
{
    return "::Ice::UnknownUserException";
}

//
// Protocol exceptions
//

const char*
Ice::ProtocolException::ice_id() const noexcept
{
    return "::Ice::ProtocolException";
}

const char*
Ice::CloseConnectionException::ice_id() const noexcept
{
    return "::Ice::CloseConnectionException";
}

const char*
Ice::DatagramLimitException::ice_id() const noexcept
{
    return "::Ice::DatagramLimitException";
}

const char*
Ice::MarshalException::ice_id() const noexcept
{
    return "::Ice::MarshalException";
}

//
// Timeout exceptions
//

const char*
Ice::TimeoutException::ice_id() const noexcept
{
    return "::Ice::TimeoutException";
}

const char*
Ice::ConnectTimeoutException::ice_id() const noexcept
{
    return "::Ice::ConnectTimeoutException";
}

const char*
Ice::CloseTimeoutException::ice_id() const noexcept
{
    return "::Ice::CloseTimeoutException";
}

const char*
Ice::InvocationTimeoutException::ice_id() const noexcept
{
    return "::Ice::InvocationTimeoutException";
}

//
// Syscall exceptions
//

Ice::SyscallException::SyscallException(
    const char* file,
    int line,
    string messagePrefix,
    SyscallException::ErrorCode error)
    : SyscallException(file, line, std::move(messagePrefix), error, IceInternal::errorToString)
{
}

Ice::SyscallException::SyscallException(
    const char* file,
    int line,
    string messagePrefix,
    ErrorCode error,
    const std::function<string(ErrorCode)>& errorToString)
    : LocalException(file, line, std::move(messagePrefix) + ": " + errorToString(error)),
      _error{error}
{
}

const char*
Ice::SyscallException::ice_id() const noexcept
{
    return "::Ice::SyscallException";
}

Ice::DNSException::DNSException(const char* file, int line, ErrorCode error, string_view host)
    : SyscallException(file, line, "cannot resolve DNS host '" + string{host} + "'", error, errorToStringDNS)
{
}

const char*
Ice::DNSException::ice_id() const noexcept
{
    return "::Ice::DNSException";
}

namespace
{
    string fileErrorToString(int error)
    {
        if (error == 0)
        {
            return "couldn't open file";
        }
        else
        {
            return IceInternal::errorToString(error);
        }
    }
}

Ice::FileException::FileException(const char* file, int line, string_view path, ErrorCode error)
    : SyscallException(file, line, "error while accessing file '" + string{path} + "'", error, fileErrorToString)
{
}

const char*
Ice::FileException::ice_id() const noexcept
{
    return "::Ice::FileException";
}

//
// Socket exceptions
//

namespace
{
    string socketErrorToString(ErrorCode error)
    {
        if (error == 0)
        {
            return "no error code";
        }
        return IceInternal::errorToString(error);
    }

    string makePrefixWithAddress(string messagePrefix, optional<string> address)
    {
        if (address.has_value() && !address->empty())
        {
            return std::move(messagePrefix) + " (remote address = " + *address + ")";
        }
        else
        {
            return messagePrefix;
        }
    }
}

Ice::SocketException::SocketException(const char* file, int line, string messagePrefix, ErrorCode error)
    : SyscallException(file, line, std::move(messagePrefix), error, socketErrorToString)
{
}

const char*
Ice::SocketException::ice_id() const noexcept
{
    return "::Ice::SocketException";
}

Ice::ConnectFailedException::ConnectFailedException(
    const char* file,
    int line,
    string messagePrefix,
    ErrorCode error,
    optional<string> address)
    : SocketException(file, line, makePrefixWithAddress(std::move(messagePrefix), std::move(address)), error)
{
}

const char*
Ice::ConnectFailedException::ice_id() const noexcept
{
    return "::Ice::ConnectFailedException";
}
namespace
{
    inline string connectionLostErrorToString(ErrorCode error)
    {
        if (error == 0)
        {
            return "no error code";
        }
        else
        {
            return socketErrorToString(error);
        }
    }
}

Ice::ConnectionLostException::ConnectionLostException(
    const char* file,
    int line,
    ErrorCode error,
    optional<string> address)
    : SocketException(
          file,
          line,
          makePrefixWithAddress("connection lost", std::move(address)),
          error,
          connectionLostErrorToString)
{
}

Ice::ConnectionLostException::ConnectionLostException(const char* file, int line, optional<string> address)
    : SocketException(file, line, makePrefixWithAddress("connection lost", std::move(address)))
{
}

const char*
Ice::ConnectionLostException::ice_id() const noexcept
{
    return "::Ice::ConnectionLostException";
}

Ice::ConnectionRefusedException::ConnectionRefusedException(const char* file, int line, optional<string> address)
    : ConnectFailedException(file, line, makePrefixWithAddress("connection refused", std::move(address)))
{
}

const char*
Ice::ConnectionRefusedException::ice_id() const noexcept
{
    return "::Ice::ConnectionRefusedException";
}

//
// Other leaf local exceptions
//

Ice::AlreadyRegisteredException::AlreadyRegisteredException(const char* file, int line, string kindOfObject, string id)
    : LocalException(file, line, "another " + kindOfObject + " is already registered with ID '" + id + "'"),
      _kindOfObject{make_shared<string>(std::move(kindOfObject))},
      _id{make_shared<string>(std::move(id))}
{
}

const char*
Ice::AlreadyRegisteredException::ice_id() const noexcept
{
    return "::Ice::AlreadyRegisteredException";
}

const char*
Ice::CommunicatorDestroyedException::ice_id() const noexcept
{
    return "::Ice::CommunicatorDestroyedException";
}

const char*
Ice::ConnectionAbortedException::ice_id() const noexcept
{
    return "::Ice::ConnectionAbortedException";
}

const char*
Ice::ConnectionClosedException::ice_id() const noexcept
{
    return "::Ice::ConnectionClosedException";
}

const char*
Ice::FeatureNotSupportedException::ice_id() const noexcept
{
    return "::Ice::FeatureNotSupportedException";
}

const char*
Ice::FixedProxyException::ice_id() const noexcept
{
    return "::Ice::FixedProxyException";
}

const char*
Ice::InitializationException::ice_id() const noexcept
{
    return "::Ice::InitializationException";
}

const char*
Ice::InvocationCanceledException::ice_id() const noexcept
{
    return "::Ice::InvocationCanceledException";
}

Ice::NoEndpointException::NoEndpointException(const char* file, int line, const ObjectPrx& proxy)
    : LocalException(file, line, "no suitable endpoint available for proxy '" + proxy.ice_toString() + "'")
{
}

const char*
Ice::NoEndpointException::ice_id() const noexcept
{
    return "::Ice::NoEndpointException";
}

Ice::NotRegisteredException::NotRegisteredException(const char* file, int line, string kindOfObject, string id)
    : LocalException(file, line, "no " + kindOfObject + " is registered with ID '" + id + "'"),
      _kindOfObject{make_shared<string>(std::move(kindOfObject))},
      _id{make_shared<string>(std::move(id))}
{
}

const char*
Ice::NotRegisteredException::ice_id() const noexcept
{
    return "::Ice::NotRegisteredException";
}

const char*
Ice::ObjectAdapterDeactivatedException::ice_id() const noexcept
{
    return "::Ice::ObjectAdapterDeactivatedException";
}

const char*
Ice::ObjectAdapterDestroyedException::ice_id() const noexcept
{
    return "::Ice::ObjectAdapterDestroyedException";
}

const char*
Ice::ObjectAdapterIdInUseException::ice_id() const noexcept
{
    return "::Ice::ObjectAdapterIdInUseException";
}

const char*
Ice::ParseException::ice_id() const noexcept
{
    return "::Ice::ParseException";
}

const char*
Ice::PluginInitializationException::ice_id() const noexcept
{
    return "::Ice::PluginInitializationException";
}

const char*
Ice::SecurityException::ice_id() const noexcept
{
    return "::Ice::SecurityException";
}

const char*
Ice::TwowayOnlyException::ice_id() const noexcept
{
    return "::Ice::TwowayOnlyException";
}

const char*
Ice::PropertyException::ice_id() const noexcept
{
    return "::Ice::PropertyException";
}

File: js/src/Ice/LocalExceptions.js
// Copyright (c) ZeroC, Inc.

import { LocalException } from "./LocalException.js";
import { identityToString } from "./IdentityToString.js";
import { ObjectPrx } from "./ObjectPrx.js";
import { Ice as Ice_Identity } from "./Identity.js";
const { Identity } = Ice_Identity;
import { Ice as Ice_ReplyStatus } from "./ReplyStatus.js";
const { ReplyStatus } = Ice_ReplyStatus;

// This file contains all the exception classes derived from LocalException defined in the Ice assembly.

//
// The 7 (8 with the RequestFailedException base class) special local exceptions that can be marshaled in an Ice reply
// message. Other local exceptions can't be marshaled.
//

export class DispatchException extends LocalException {
    static get _ice_id() {
        return "::Ice::DispatchException";
    }

    constructor(replyStatus, message) {
        super(DispatchException.createDispatchFailedMessage(replyStatus, message));
        this._replyStatus = replyStatus;

        DEV: console.assert(
            replyStatus.value > ReplyStatus.UserException.value && replyStatus.value <= 255,
            "The reply status must fit in a byte and have a value greater than UserException",
        );
    }

    get replyStatus() {
        return this._replyStatus;
    }

    static createDispatchFailedMessage(replyStatus, message) {
        if (message !== undefined) {
            return message;
        } else {
            return `The dispatch failed with reply status ${replyStatus}.`;
        }
    }
}

export class RequestFailedException extends DispatchException {
    static get _ice_id() {
        return "::Ice::RequestFailedException";
    }

    constructor(replyStatus, id = new Identity(), facet = "", operation = "") {
        super(replyStatus, RequestFailedException.createRequestFailedMessage(replyStatus, id, facet, operation));
        this._id = id;
        this._facet = facet;
        this._operation = operation;
    }

    get id() {
        return this._id;
    }

    get facet() {
        return this._facet;
    }

    get operation() {
        return this._operation;
    }

    static createRequestFailedMessage(replyStatus, id, facet, operation) {
        if (id.name.length > 0) {
            return `Dispatch failed with ${replyStatus} { id = '${identityToString(id)}', facet = '${facet}', operation = '${operation}' }`;
        } else {
            return undefined;
        }
    }
}

export class ObjectNotExistException extends RequestFailedException {
    static get _ice_id() {
        return "::Ice::ObjectNotExistException";
    }

    constructor(id, facet, operation) {
        super(ReplyStatus.ObjectNotExist, id, facet, operation);
    }
}

export class FacetNotExistException extends RequestFailedException {
    static get _ice_id() {
        return "::Ice::FacetNotExistException";
    }

    constructor(id, facet, operation) {
        super(ReplyStatus.FacetNotExist, id, facet, operation);
    }
}

export class OperationNotExistException extends RequestFailedException {
    static get _ice_id() {
        return "::Ice::OperationNotExistException";
    }

    constructor(id, facet, operation) {
        super(ReplyStatus.OperationNotExist, id, facet, operation);
    }
}

export class UnknownException extends DispatchException {
    static get _ice_id() {
        return "::Ice::UnknownException";
    }

    constructor(message, replyStatus = ReplyStatus.UnknownException) {
        super(replyStatus, message);
    }

    get unknown() {
        return this.message;
    }
}

export class UnknownLocalException extends UnknownException {
    static get _ice_id() {
        return "::Ice::UnknownLocalException";
    }

    constructor(message) {
        super(message, ReplyStatus.UnknownLocalException);
    }
}

export class UnknownUserException extends UnknownException {
    static get _ice_id() {
        return "::Ice::UnknownUserException";
    }

    constructor(message) {
        super(message, ReplyStatus.UnknownUserException);
    }
}

//
// Protocol exceptions
//

export class ProtocolException extends LocalException {
    static get _ice_id() {
        return "::Ice::ProtocolException";
    }
}

export class CloseConnectionException extends ProtocolException {
    static get _ice_id() {
        return "::Ice::CloseConnectionException";
    }

    constructor() {
        super("Connection closed by the peer.");
    }
}

export class MarshalException extends ProtocolException {
    static get _ice_id() {
        return "::Ice::MarshalException";
    }
}

//
// Timeout exceptions
//

export class TimeoutException extends LocalException {
    static get _ice_id() {
        return "::Ice::TimeoutException";
    }

    constructor(message = undefined, ...params) {
        super(message || "operation timed out", ...params);
    }
}

export class CloseTimeoutException extends TimeoutException {
    static get _ice_id() {
        return "::Ice::CloseTimeoutException";
    }

    constructor() {
        super("Close timed out.");
    }
}

export class ConnectTimeoutException extends TimeoutException {
    static get _ice_id() {
        return "::Ice::ConnectTimeoutException";
    }

    constructor() {
        super("Connect timed out.");
    }
}

export class InvocationTimeoutException extends TimeoutException {
    static get _ice_id() {
        return "::Ice::InvocationTimeoutException";
    }

    constructor() {
        super("Invocation timed out.");
    }
}

//
// Syscall exceptions
//

export class SyscallException extends LocalException {
    static get _ice_id() {
        return "::Ice::SyscallException";
    }
}

//
// Socket exceptions
//

export class SocketException extends SyscallException {
    static get _ice_id() {
        return "::Ice::SocketException";
    }
}

export class ConnectFailedException extends SocketException {
    static get _ice_id() {
        return "::Ice::ConnectFailedException";
    }
}

export class ConnectionLostException extends SocketException {
    static get _ice_id() {
        return "::Ice::ConnectionLostException";
    }
}

export class ConnectionRefusedException extends ConnectFailedException {
    static get _ice_id() {
        return "::Ice::ConnectionRefusedException";
    }
}

//
// Other leaf local exceptions in alphabetical order.
//

export class AlreadyRegisteredException extends LocalException {
    static get _ice_id() {
        return "::Ice::AlreadyRegisteredException";
    }

    constructor(kindOfObject, id) {
        super(`Another ${kindOfObject} is already registered with ID '${id}'.`);
        this._kindOfObject = kindOfObject;
        this._id = id;
    }

    get kindOfObject() {
        return this._kindOfObject;
    }

    get id() {
        return this._id;
    }
}

export class CommunicatorDestroyedException extends LocalException {
    static get _ice_id() {
        return "::Ice::CommunicatorDestroyedException";
    }

    constructor() {
        super("Communicator destroyed.");
    }
}

export class ConnectionAbortedException extends LocalException {
    static get _ice_id() {
        return "::Ice::ConnectionAbortedException";
    }

    constructor(message, closedByApplication) {
        super(message);
        this.closedByApplication = closedByApplication;
    }
}

export class ConnectionClosedException extends LocalException {
    static get _ice_id() {
        return "::Ice::ConnectionClosedException";
    }

    constructor(message, closedByApplication) {
        super(message);
        this.closedByApplication = closedByApplication;
    }
}

export class FeatureNotSupportedException extends LocalException {
    static get _ice_id() {
        return "::Ice::FeatureNotSupportedException";
    }

    constructor(message) {
        super(message);
    }
}

export class FixedProxyException extends LocalException {
    static get _ice_id() {
        return "::Ice::FixedProxyException";
    }

    constructor() {
        super("Cannot change the connection properties of a fixed proxy.");
    }
}

export class InitializationException extends LocalException {
    static get _ice_id() {
        return "::Ice::InitializationException";
    }
}

export class InvocationCanceledException extends LocalException {
    static get _ice_id() {
        return "::Ice::InvocationCanceledException";
    }

    constructor() {
        super("Invocation canceled.");
    }
}

export class NoEndpointException extends LocalException {
    static get _ice_id() {
        return "::Ice::NoEndpointException";
    }

    constructor(messageOrProxy) {
        super(NoEndpointException.createMessage(messageOrProxy));
    }

    static createMessage(arg) {
        return arg instanceof ObjectPrx ? `$"No suitable endpoint available for proxy '${arg}'.` : arg;
    }
}

export class NotRegisteredException extends LocalException {
    static get _ice_id() {
        return "::Ice::NotRegisteredException";
    }

    constructor(kindOfObject, id) {
        super(`No ${kindOfObject} is registered with ID '${id}'.`);
        this._kindOfObject = kindOfObject;
        this._id = id;
    }

    get kindOfObject() {
        return this._kindOfObject;
    }

    get id() {
        return this._id;
    }
}

export class ObjectAdapterDestroyedException extends LocalException {
    static get _ice_id() {
        return "::Ice::ObjectAdapterDestroyedException";
    }

    constructor(name) {
        super(`Object adapter '${name}' is destroyed.`);
    }
}

export class ParseException extends LocalException {
    static get _ice_id() {
        return "::Ice::ParseException";
    }
}

export class TwowayOnlyException extends LocalException {
    static get _ice_id() {
        return "::Ice::TwowayOnlyException";
    }

    constructor(operation) {
        super(`Cannot invoke operation '${operation}' with a oneway, batchOneway, datagram, or batchDatagram proxy.`);
        this.operation = operation;
    }
}

export class PropertyException extends LocalException {
    static get _ice_id() {
        return "::Ice::PropertyException";
    }

    constructor(message) {
        super(message);
    }
}
